### ISSUE TITLE
Slicer / module .so troubles

### ISSUE THREAD

**Issue by DESm1th** [TIGRLab/admin#1291]
Rutwik uncovered a few issues with slicer:

1. When slicer 4.4 is loaded and then FSL 5.0.7 is loaded FSL resets the LD_LIBRARY_PATH and prevents slicer from finding its .so files
2. Sometimes even when the LD_LIBRARY_PATH is set correctly for slicer it still crashes looking for .so files that are in the slicer build folder (and are listed among the LD_LIBRARY_PATHs)
3. Additionally the slicer modifications to the LD_LIBRARY_PATH are permanently damaging the shell environment (has to be closed and a new one opened to fix it), breaking the ability to use ssh and possibly other commands as well.

Need a better way to handle slicer :( Possibly a container, or a wrapper script for the binary that temporarily sets LD_LIBRARY_PATH. Will do more research for the best option.

**Comment by DESm1th** [comment]
After a little further investigation into problem #1 it seems that 'prepend-path' with LD_LIBRARY_PATH just overwrites the whole path rather than prepending it. It definitely worked in the past, and still works on machines that have not had the fix from https://github.com/TIGRLab/admin/issues/1280 applied to them.

**Comment by DESm1th** [comment]
Just an update: 

The reason slicer breaks the module commands without the fix at #1280 is because modulecmd relies on /usr/lib/x86_64-linux-gnu/libtcl8.6.so and slicer adds /opt/quarantine/slicer/4.4.0/build/lib/TclTk/lib/libtcl8.6.so to the LD_LIBRARY_PATH

Additionally, it may be the case that the same fix breaks the 'prepend-path' directive for LD_LIBRARY_PATH because, according to the man page for ld.so, having a setuid or setguid bit set on an executable can turn on 'secure-execution mode' for the dynamic linker and change how the OS manages certain environment variables like LD_LIBRARY_PATH

**Comment by DESm1th** [comment]
Ok, after reading a bit the temporary fix will be to remove any libtcl8.6.so files in any of the slicer directories and replace them with a link to the operating system version. I'll store these problem versions in a folder inside the same directory with a README file just in case they're ever needed again. Any machine that had the fix from #1280 will also need that fix rolled back to avoid the issues with LD_LIBRARY_PATH inconsistencies.

This doesn't really address the main problem: A lot of our module files modify LD_LIBRARY_PATH and that's bad practice (for a lot of reasons described [here](http://www.hpc.dtu.dk/?page_id=1180)). I'll keep an eye open for better long term solutions but we may be stuck :(

**Comment by josephdviviano** [comment]
@DESm1th -- what exactly is the issue with #1280 when not using slicer? aside from changing the LD_LIBRARY_PATH being bad practice, as I understand it, the only software that causes issues is slicer?

**Comment by DESm1th** [comment]
#1280 affects all modules that make modifications to LD_LIBRARY_PATH, which seems to be a few of them. Having setuid or setguid bits set breaks the prepend-path directive when working with that environment variable. 

An example:

On a computer that has had that fix applied, load some module that modifies LD_LIBRARY_PATH and then echo LD_LIBRARY_PATH. Then load a second module that also modifies it and echo the path again. Instead of seeing:

LD_LIBRARY_PATH=module_two_changes;module_one_changes

You see:

LD_LIBRARY_PATH=module_two_changes

Only the last module loaded will have the variable set as it requires. The 'fix' of just linking back to the system shared object file lets slicer work without needing fix #1280 so we can avoid that but it doesn't address any of the other negative effects of setting LD_LIBRARY_PATH, so for instance if you load slicer and try to ssh somewhere you'll get error messages regarding openssl.so because slicer has it's own copy of that and it's not configured correctly for ssh. We'll keep hitting little snags like this unless we can come up with a better fix or a better way to handle our shared object files for our modules :(

But I'm not sure what to do short of recompiling everything (may not be possible), tediously wrapping every binary that's affected in a shell script that handles the modifications, or making containers for everything so they don't interact with each other.

**Comment by josephdviviano** [comment]
@DESm1th -- this must only affect a small number of module configurations or we would have noticed it ages ago.

Crucially, we really shouldn't be setting LD_LIBRARY_PATH in modules at all, though in practice it hasn't mattered much.

Why don't we make these conflicting modules mutually exclusive? Without #1280, we can't load modules at all, so that's not a good solution. However, it's trivial to add `module loads` and `module unloads` as part of the BASH script, which gets around the problem?

**Comment by DESm1th** [comment]
Hey Joe, 

The fix #1280 was only applied recently by you to a few computers and by me yesterday to mansfield and the machine rutwik uses so we wouldn't have had much opportunity to notice. 

We can load modules without the fix, because I replaced the broken slicer tcl.so file with a link to the working one that's installed on each machine. Modules work fine now, the problem is we'll still have other side effects because the root cause (modifying LD_LIBRARY_PATH) has not been addressed.

The affected modules include, at a glance, include slicer, FSL, R, minc-toolkit, freesurfer/5.3.0 and freesurfer/6.0.0, and CIVET. In other words, if we made such modules mutually exclusive to continue using fix #1280 a lot of our datman scripts wouldnt be able to run because they depend on multiple of these at the same time.
